src/gui/tray/UserModel.cpp

```diff
@@ -105,8 +105,71 @@ void User::slotBuildNotificationDisplay(const ActivityList &list)
 {
-    qCDebug(lcActivity) << "Starting Notification refresh timer with " << interval.count() / 1000 << " sec interval";
-    _notificationCheckTimer.start(interval.count());
+    if (!checkPushNotificationsAndActivitiesAreReady()) {
+        qCDebug(lcActivity) << "Starting Notification refresh timer with " << interval.count() / 1000 << " sec interval";
+        _notificationCheckTimer.start(interval.count());
+    }
+}
+
+void User::slotPushNotificationsReady()
+{
+    qCInfo(lcActivity) << "Push notifications are ready";
+
+    if (_notificationCheckTimer.isActive()) {
+        // as we are now able to use push notifications - let's stop the polling timer
+        _notificationCheckTimer.stop();
+    }
+
+    onPushNotificationsReady();
+}
+
+void User::slotPushNotificationsDisabled()
+{
+    disconnect(_account->account()->pushNotifications(), &PushNotifications::notificationsChanged, this, &User::slotReceivedPushNotification);
+    disconnect(_account->account()->pushNotifications(), &PushNotifications::activitiesChanged, this, &User::slotReceivedPushActivity);
+
+    disconnect(_account->account().data(), &Account::pushNotificationsDisabled, this, &User::slotPushNotificationsDisabled);
+
+    // connection to WebSocket may have dropped or an error occured, so we need to bring back the polling until we have re-established the connection
+    ConfigFile cfg;
+    setNotificationRefreshInterval(cfg.notificationRefreshInterval());
+}
+
+void User::slotReceivedPushNotification(Account *account)
+{
+    if (account->id() == _account->account()->id()) {
+        this->slotRefreshNotifications();
+    }
+}
+
+void User::slotReceivedPushActivity(Account *account)
+{
+    if (account->id() == _account->account()->id()) {
+        this->slotRefreshActivities();
+    }
+}
+void User::onPushNotificationsReady() const
+{
+    connect(_account->account().data(), &Account::pushNotificationsDisabled, this, &User::slotPushNotificationsDisabled, Qt::UniqueConnection);
+
+    connect(_account->account()->pushNotifications(), &PushNotifications::notificationsChanged, this, &User::slotReceivedPushNotification, Qt::UniqueConnection);
+    connect(_account->account()->pushNotifications(), &PushNotifications::activitiesChanged, this, &User::slotReceivedPushActivity, Qt::UniqueConnection);
+}
+
+bool User::checkPushNotificationsAndActivitiesAreReady() const
+{
+    const auto pushNotifications = _account->account()->pushNotifications();
+
+    const auto pushActivitiesAvailable = _account->account()->capabilities().availablePushNotifications() & PushNotificationType::Activities;
+    const auto pushNotificationsAvailable = _account->account()->capabilities().availablePushNotifications() & PushNotificationType::Notifications;
+
+    if ((pushActivitiesAvailable && pushNotificationsAvailable) && pushNotifications && pushNotifications->isReady()) {
+        onPushNotificationsReady();
+        return true;
+    } else {
```

Comment:
I think `else` could be omitted but that's a matter of taste.

Reply:
Indeed, matter of taste. ;-)

When both branches are balanced like that I don't mind either way.

Reply:
@er-vin Yep. But, in this case, when if/else is not big, I'd prefer keeping it here. Plus, it makes it clear when someone will want to change something here the code should go either into if or else. Omitting else would bring more freedom for error IMHO. Don't know if my explanation is clear, I am not sure myself the best way to explain this choice :)

Reply:
@FlexW Good catch. Since I've removed the enum, it's not anymore.